osbs/api.py

```diff
@@ -189,19 +191,57 @@ def _panic_msg_for_more_running_builds(self, build_config_name, builds):
             (build_config_name, builds)
         return msg
 
-    def _create_build_config_and_build(self, build_request):
-        # TODO: test this method more thoroughly
-        build_json = build_request.render()
-        api_version = build_json['apiVersion']
-        if api_version != self.os_conf.get_openshift_api_version():
-            raise OsbsValidationException("BuildConfig template has incorrect apiVersion (%s)" %
-                                          api_version)
+    def _verify_labels_match(self, new_build_config, existing_build_config):
+        new_labels = new_build_config['metadata']['labels']
+        existing_labels = existing_build_config['metadata']['labels']
 
-        build_config_name = build_json['metadata']['name']
+        for key in self._GIT_LABEL_KEYS:
+            new_label_value = new_labels.get(key)
+            existing_label_value = existing_labels.get(key)
+
+            if (existing_label_value is not None and
+                    existing_label_value != new_label_value):
+
+                msg = (
+                    'Git labels collide with existing build config labels. '
+                    'Existing labels: %r, '
+                    'New labels: %r ') % (existing_labels, new_labels)
+                raise OsbsValidationException(msg)
+
+    def _get_existing_build_config(self, build_config):
+        """
+        Uses the given build config to find an existing matching build config.
+        Build configs are a match if:
+        - metadata.name are equal
+        OR
+        - metadata.labels.git-repo-name AND metadata.labels.git-branch are equal
+        """
+
+        git_labels = [(key, build_config['metadata']['labels'][key])
+                      for key in self._GIT_LABEL_KEYS]
+        name = build_config['metadata']['name']
+
+        queries = (
+            (self.os.get_build_config_by_labels, git_labels),
+            (self.os.get_build_config, name),
+        )
+
+        existing_bc = None
+        for func, arg in queries:
+            try:
+                existing_bc = func(arg)
+                # build config found
+                break
+            except OsbsException:
```

Comment:
I was thinking of using new exception, something like `OsbsBuildConfigNotFound`, but now I see we must also handle 404 (or something) from `get_build_config`. At least logging the exception sounds reasonable.